compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java

```diff
@@ -0,0 +1,1129 @@
+
+        for (BType inferableType : inferableTypes) {
+            for (BType expectedType : expectedTypes) {
+                if (inferableType.tag == expectedType.tag) {
+                    matchedTypes.add(expectedType);
+                }
+            }
+        }
+
+        if (matchedTypes.size() == 1) {
+            return matchedTypes.iterator().next();
+        }
+
+        return null;
+    }
+
+    private boolean refersInferableParamName(List<String> paramsWithInferredTypedescDefault, BType type) {
+        return refersInferableParamName(paramsWithInferredTypedescDefault, type, new HashSet<>());
+    }
+
+    private boolean refersInferableParamName(List<String> paramsWithInferredTypedescDefault, BType type,
+                                             Set<BType> unresolvedTypes) {
+        if (!unresolvedTypes.add(type)) {
+            return false;
+        }
+
+        switch (type.tag) {
+            case TypeTags.PARAMETERIZED_TYPE:
+                String paramName = ((BParameterizedType) type).paramSymbol.name.value;
+                return paramsWithInferredTypedescDefault.contains(paramName);
+            case TypeTags.XML:
+                return refersInferableParamName(paramsWithInferredTypedescDefault, ((BXMLType) type).constraint,
+                                                unresolvedTypes);
+            case TypeTags.TABLE:
+                BTableType tableType = (BTableType) type;
+                if (refersInferableParamName(paramsWithInferredTypedescDefault, tableType.constraint,
+                                             unresolvedTypes)) {
+                    return true;
+                }
+
+                BType keyTypeConstraint = tableType.keyTypeConstraint;
+                if (keyTypeConstraint == null) {
+                    return false;
+                }
+
+                return refersInferableParamName(paramsWithInferredTypedescDefault, keyTypeConstraint, unresolvedTypes);
+            case TypeTags.TYPEDESC:
+                return refersInferableParamName(paramsWithInferredTypedescDefault, ((BTypedescType) type).constraint,
+                                                unresolvedTypes);
+            case TypeTags.STREAM:
+                BStreamType streamType = (BStreamType) type;
+                if (refersInferableParamName(paramsWithInferredTypedescDefault, streamType.constraint,
+                                             unresolvedTypes)) {
+                    return true;
+                }
+
+                BType streamErrorType = streamType.error;
+                if (streamErrorType == null) {
+                    return false;
+                }
+                return refersInferableParamName(paramsWithInferredTypedescDefault, streamErrorType, unresolvedTypes);
+            case TypeTags.INVOKABLE:
+//                if (Symbols.isFlagOn(type.flags, Flags.ANY_FUNCTION)) {
+//                    return false;
+//                }
```

Comment:
Intentionally left out?

Reply:
I added this bit before the `function` type descriptor change was merged to the master branch but missed un-commenting it :sweat_smile: Fixed and added tests, thanks!